lib/obj/src/function.rs

```diff
@@ -0,0 +1,75 @@
+use cranelift_codegen::settings;
+use cranelift_codegen::settings::Configurable;
+use cranelift_entity::EntityRef;
+use faerie::{Artifact, Decl, Link, RelocOverride};
+use wasmtime_environ::{Compilation, Module, RelocationTarget, Relocations};
+
+/// Defines module functions
+pub fn declare_functions(
+    obj: &mut Artifact,
+    module: &Module,
+    relocations: &Relocations,
+) -> Result<(), String> {
+    for (i, _function_relocs) in relocations.iter().rev() {
+        let func_index = module.func_index(i);
+        let string_name = format!("_wasm_function_{}", func_index.index());
+        obj.declare(string_name, Decl::Function { global: true })
+            .map_err(|err| format!("{}", err))?;
+    }
+    Ok(())
+}
+
+/// Emits module functions
+pub fn emit_functions(
+    obj: &mut Artifact,
+    module: &Module,
+    compilation: &Compilation,
+    relocations: &Relocations,
+) -> Result<(), String> {
+    debug_assert!(
+        module.start_func.is_none()
+            || module.start_func.unwrap().index() >= module.imported_funcs.len(),
+        "imported start functions not supported yet"
+    );
+
+    let mut shared_builder = settings::builder();
+    shared_builder
+        .enable("enable_verifier")
+        .expect("Missing enable_verifier setting");
+
+    for (i, _function_relocs) in relocations.iter() {
+        let body = &compilation.functions[i];
+        let func_index = module.func_index(i);
+        let string_name = format!("_wasm_function_{}", func_index.index());
+
+        obj.define(string_name, body.clone())
+            .map_err(|err| format!("{}", err))?;
+    }
+
+    for (i, function_relocs) in relocations.iter() {
+        let func_index = module.func_index(i);
+        let string_name = format!("_wasm_function_{}", func_index.index());
+        for r in function_relocs {
+            match r.reloc_target {
+                RelocationTarget::UserFunc(target_index) => {
+                    let target_name = format!("_wasm_function_{}", target_index.index());
+                    obj.link_with(
+                        Link {
+                            from: &string_name,
+                            to: &target_name,
+                            at: r.offset as u64,
+                        },
+                        RelocOverride {
+                            reloc: 0,
```

Comment:
The relocation here probably can’t be 0

Reply:
Not sure what it needs to be and doc is not clear there either.

Reply:
If you’re getting relocations from cranelift it’s probably in the `r` strict is my guess.
Or did you mean faerie docs aren’t clear.
Regardless a relocation of 0 will almost certainly not be correct and will likely cause either a link time error (or worse a runtime segfault).
I regret having added the link_with api because was worried it might be used like this :(
Anyway my guess is you have a relocation value somewhere in r?

Reply:
@m4b can you explain what the `reloc` field mean? 0 worked in my example.

The [cranelift struct](https://github.com/CraneStation/cranelift/blob/4b698c2b3b5ec4702a5494c6636a723d5b6df466/lib/simplejit/src/backend.rs#L100) has `offset`, reloc type, `name` and `addend` fields. The `addend` field in the `RelocOverride` is clear, but it is not clear what is the purpose of `reloc` for me.

Reply:
You can copy https://github.com/CraneStation/cranelift/blob/4b698c2b3b5ec4702a5494c6636a723d5b6df466/lib/faerie/src/container.rs#L15-L65 here and use it.

Reply:
I decided to removing support for `r.addend != 0` for now -- works without it for my subproject